### PR TITLE
ISTIO support for notebook controller

### PR DESCRIPTION
**Pull request by gabrielwen**
Relates to #2978

Tested through manual deployment:
1. Remove ambassador from component list.
1. Bring up KF cluster.
1. Navigate to `Notebooks` and use the UI to create a new notebook server.
1. Verify that `CONNECT` could correctly find Jupyter notebook UI page.
1. Verify that when deleting a notebook server, virtual service is gone with it.

/assign @lluunn

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/3104)

### CODE REVIEW

File: components/notebook-controller/pkg/controller/notebook/notebook_controller.go [hunk 1]
```diff
@@ -204,6 +214,12 @@ func (r *ReconcileNotebook) Reconcile(request reconcile.Request) (reconcile.Resu
 		if err != nil {
 			return reconcile.Result{}, err
 		}
+		log.Info("Creating virtual service", "namespace", service.Namespace, "name",
```
Comment (lluunn) [inline components/notebook-controller/pkg/controller/notebook/notebook_controller.go]: shouldn't you do another check for virtualService by itself?
This if block is when you don't see the "service"?

File: kubeflow/jupyter/notebook_controller.libsonnet [hunk 2]
```diff
@@ -1,32 +1,32 @@
 {
-  local util = import "kubeflow/common/util.libsonnet",
```
Comment (lluunn) [inline kubeflow/jupyter/notebook_controller.libsonnet]: we use double quote else where.
I think maybe your editor/formatter did this?
Could you revert the change here

**Comment by lluunn** [comment]
Looks great.
Have you test this (create/ delete notebook, check controller logs)?
After testing, could you add it (what steps you did, what's the result) to PR description?
Thanks.

File: components/notebook-controller/pkg/controller/notebook/notebook_controller.go (hunk 1 above)
Comment (gabrielwen) [reply components/notebook-controller/pkg/controller/notebook/notebook_controller.go]: done.

File: kubeflow/jupyter/notebook_controller.libsonnet (hunk 2 above)
Comment (gabrielwen) [reply kubeflow/jupyter/notebook_controller.libsonnet]: yea... it's `vim-jsonnet`...  reverted.

File: components/notebook-controller/pkg/controller/notebook/notebook_controller.go [hunk 3]
```diff
@@ -216,6 +219,28 @@ func (r *ReconcileNotebook) Reconcile(request reconcile.Request) (reconcile.Resu
...
+	if err != nil && errors.IsNotFound(err) {
+		log.Info("Creating virtual service", "namespace", instance.Namespace, "name",
+			virtualServiceName(instance.Name, instance.Namespace))
+		err = r.Create(context.TODO(), virtualService)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else if err != nil {
+		return reconcile.Result{}, err
+	}
```
Comment (lluunn) [inline components/notebook-controller/pkg/controller/notebook/notebook_controller.go]: you need an update step here, see 213~220 for service, or 185~193 for ss

Comment (gabrielwen) [reply components/notebook-controller/pkg/controller/notebook/notebook_controller.go]: doesn't seem to have a proper way of doing it - having this error when trying to run update:
`"error":"virtualservices.networking.istio.io \"notebook-kubeflow-nb-testing\" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update`

Comment (gabrielwen) [reply components/notebook-controller/pkg/controller/notebook/notebook_controller.go]: added.

File: components/notebook-controller/pkg/util/util.go [hunk 4]
```diff
@@ -79,3 +80,27 @@ func CopyServiceFields(from, to *corev1.Service) bool {
...
+	if !found {
+		return true
+	}
+	if err != nil {
+		return true
+	}
+
+	toSpec, found, err := unstructured.NestedMap(to.Object, "spec")
+	if !found {
+		return true
```
Comment (lluunn) [inline components/notebook-controller/pkg/util/util.go]: we should not return here.
We should still SetNestedMap to copy spec from "from" to "to" right?

Comment (gabrielwen) [reply components/notebook-controller/pkg/util/util.go]: done.

**Comment by lluunn** [comment]
thank you! 
/lgtm
/approve

/hold
Could you test one more thing:
remove/edit the virtualService and make sure the reconcile recover the virtualService.
If that works, please cancel the hold

**Comment by gabrielwen** [comment]
/hold cancel

watch event added and verified.

**Comment by lluunn** [comment]
nice!
/lgtm

**Comment by gabrielwen** [comment]
Fixes #2172
